FreeRTOS/Test/CMock/queue/semaphore/counting_semaphore_utest.c

```diff
@@ -0,0 +1,557 @@
+ * xSemaphoreCreateCounting.
+ * @coverage xQueueGenericSend
+ */
+void test_macro_xSemaphoreGive_CountingSemaphore_one_zero_success( void )
+{
+    SemaphoreHandle_t xSemaphore = xSemaphoreCreateCounting( 1, 0 );
+
+    /* validate returned semaphore handle */
+    TEST_ASSERT_NOT_EQUAL( NULL, xSemaphore );
+
+    TEST_ASSERT_EQUAL( QUEUE_T_SIZE, getLastMallocSize() );
+
+    /* Verify that an xSemaphoreGive succeeds */
+    TEST_ASSERT_EQUAL( pdTRUE, xSemaphoreGive( xSemaphore ) );
+
+    vSemaphoreDelete( xSemaphore );
+}
+
+/**
+ * @brief Test xSemaphoreGive and xSemaphoreTake with xSemaphoreCreateCounting( 1, 0 )
+ * @details Test xSemaphoreGive and xSemaphoreTake with a binary semaphore
+ * constructed with xSemaphoreCreateCounting.
+ * @coverage xQueueGenericSend xQueueSemaphoreTake
+ */
+void test_macro_xSemaphoreGive_xSemaphoreTake_CountingSemaphore_one_zero_success( void )
+{
+    SemaphoreHandle_t xSemaphore = xSemaphoreCreateCounting( 1, 0 );
+
+    /* validate returned semaphore handle */
+    TEST_ASSERT_NOT_EQUAL( NULL, xSemaphore );
+
+    TEST_ASSERT_EQUAL( QUEUE_T_SIZE, getLastMallocSize() );
+
+    /* Verify that an xSemaphoreGive succeeds */
+    TEST_ASSERT_EQUAL( pdTRUE, xSemaphoreGive( xSemaphore ) );
+
+    /* Verify that an xSemaphoreTake succeeds */
+    TEST_ASSERT_EQUAL( pdTRUE, xSemaphoreTake( xSemaphore, 0 ) );
+
+    /* Verify that an xSemaphoreGive succeeds */
+    TEST_ASSERT_EQUAL( pdTRUE, xSemaphoreGive( xSemaphore ) );
+
+    vSemaphoreDelete( xSemaphore );
+}
+
+
+/**
+ * @brief Test xSemaphoreGive with xSemaphoreCreateCounting( 100, 50 )
+ * @details Test multiple xSemaphoreGive calls on a counting semaphore
+ * @coverage xQueueGenericSend
+ */
+void test_macro_xSemaphoreGive_CountingSemaphore_100_50( void )
+{
+    SemaphoreHandle_t xSemaphore = xSemaphoreCreateCounting( 100, 50 );
+
+    /* validate returned semaphore handle */
+    TEST_ASSERT_NOT_EQUAL( NULL, xSemaphore );
+
+    TEST_ASSERT_EQUAL( QUEUE_T_SIZE, getLastMallocSize() );
+
+    /* Call xSemaphoreGive 50 times and verify success */
+    for( int i = 0; i < 50; i++ )
+    {
+        /* Verify that an xSemaphoreGive operation succeeds */
+        TEST_ASSERT_EQUAL( pdTRUE, xSemaphoreGive( xSemaphore ) );
+    }
+
+    /* Check the count */
+    TEST_ASSERT_EQUAL( 100, uxSemaphoreGetCount( xSemaphore ) );
+
+    /* Veirfy that a subsequent call to xSemaphoreGive fails */
+    TEST_ASSERT_EQUAL( pdFALSE, xSemaphoreGive( xSemaphore ) );
+
+    /* Verify that an xSemaphoreTake operation succeeds */
+    TEST_ASSERT_EQUAL( pdTRUE, xSemaphoreTake( xSemaphore, 0 ) );
+
+    /* Check the count */
+    TEST_ASSERT_EQUAL( 99, uxSemaphoreGetCount( xSemaphore ) );
+
+    vSemaphoreDelete( xSemaphore );
+}
+
+/**
+ * @brief Test xSemaphoreTake with xSemaphoreCreateCounting( 100, 50 )
+ * @details Test multiple xSemaphoreTake calls on a counting semaphore
+ * @coverage xQueueSemaphoreTake
+ */
+void test_macro_xSemaphoreTake_CountingSemaphore_100_50( void )
```

Comment:
Should there be tests which verifies blocking with counting semaphore ?

Reply:
Added.